cirq/protocols/json.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from typing import Union, Any, Dict, Optional, List, Callable, Type, cast
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+
+class _ResolverCache:
+    """Lazily import and build registry to avoid circular imports."""
+
+    def __init__(self):
+        self._crd = None
+
+    @property
+    def cirq_class_resolver_dictionary(self) -> Dict[str, Type]:
+        if self._crd is None:
+            import cirq
+            import cirq.ops.pauli_gates  # for mypy
+            import cirq.devices.unconstrained_device  # for mypy
+            self._crd = {
```

Comment:
Probably a bad idea, but we could use `inspect`'s getmembers to construct this?

Reply:
If you look through the history of commits, I used to just `import cirq; getattr(cirq, cirq_type_str)`. @Strilanc and I agreed that this was too open-ended